### PR TITLE
Add Zenoh pico support

### PR DESCRIPTION
**Pull request by PetervdPerk-NXP**
Zenoh is a publish/subscribe protocol supporting peer-to-peer (FMU-to-FMU) and routed communication (FUM-to-Companion/Linux). Zenoh is plug-gable and on Linux it can be bridged to ROS2/DDS through [zenoh-plugin-dds](https://github.com/eclipse-zenoh/zenoh-plugin-dds).

Zenoh-Pico is an implementation of Zenoh in C for constrained devices, it's fully compatible with the [Rust Zenoh implementation](https://github.com/eclipse-zenoh/zenoh).

For serialization of the messages the CycloneDDS CDRSTREAM library is used to generate DDS/ROS2 byte compatible serialization. This is decoupled from Zenoh so that other modules can re-use the serialization library.

# Setup (FMU-to-Companion/Linux)
Companion/Linux
- Setup Zenoh version on a Linux system https://github.com/eclipse-zenoh/zenoh#how-to-install-it
  - Tested with zenohd v0.7.0-rc-401-gb9103c3d and zenohd v0.10.0-dev-23-g44875300 (Needed for multicast support)
- Setup Zenoh-plugin-dds on the Linux system https://github.com/eclipse-zenoh/zenoh-plugin-dds#how-to-install-it  
  - Tested with zenoh-bridge-dds v0.7.2-rc built with rustc 1.70.0 (90c541806 2023-05-31)
- Start a Zenoh router on the  Linux machine connected to your FMU 
  - For client mode use `zenohd`
  - For multicast mode use `zenohd -e udp/224.0.0.225:7447` note zenohd v0.10.0-dev-23-g44875300 or newer required
- Start Zenoh dds bridge on the Linux machine `zenoh-bridge-dds -m client `

FMU
- Enable the Zenoh module on your FMU build `CONFIG_DRIVERS_ZENOH`. 
- Add a uORB topic to published i.e.  `zenoh config addpublisher sensor_combined sensor`
- Start Zenoh `zenoh start`

Companion/Linux
 - Make sure you've got ROS2 installed (in my case I've been testing ROS2 Galactic but newer should also work)
 - Clone https://github.com/PX4/px4_msgs and build px4_msgs
 - Source ROS2 and compiled px4_msgs
 - Type `ros2 topic echo sensor px4_msgs/SensorCombined` to receive msgs from the FMU
```
timestamp: 5101022267
gyro_rad:
- 0.00033584635821171105
- 0.001161606633104384
- 0.0029726973734796047
gyro_integral_dt: 4390
accelerometer_timestamp_relative: 0
accelerometer_m_s2:
- -0.1838383674621582
- -0.04610908403992653
- -9.762420654296875
accelerometer_integral_dt: 4390
accelerometer_clipping: 0
gyro_clipping: 1
accel_calibration_count: 1
gyro_calibration_count: 1

```

# Current shortcomings

- Naive CSV file method for storing configuration and publishers/subscribers, would be nice PX4 param system can be expanded to support custom key/value storage i.e. json/xml like.
- No message checking i.e. different versions of PX4 might produce corrupted data
- No rate limiting yet
- No timesync
- FMU-to-FMU communication no distinction between own msgs and from incoming from Zenoh
- Zenoh 0.7.X based protocol, new 0.10.0 upcoming but probably stabilize around Q1 2024
- Documentation, I would consider Zenoh for now to experimental when all stabalizes I think we could add documentation.

### CODE REVIEW

**Comment by dagar** [comment]
This will take a little bit to review, but in the meantime can you get it through CI cleanly?

Screenshot 2023-08-29 at 10 21 51 AM

**Comment by PetervdPerk-NXP** [comment]
Hm to get Zenoh-pico to build, I think CMake 1.14 is mandatory need to check if CMake 1.13 might work as well.

File: src/drivers/zenoh/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+endfunction()
+
+set(POSIX_COMPATIBLE YES)
+set(BUILD_SHARED_LIBS OFF)
+set(BUILD_TESTING OFF)
+set(CHECK_THREADS NO)
+set(MESSAGE_QUIET ON)
+#set(ZENOH_DEBUG 3)
+
+add_subdirectory(zenoh-pico)
```
Comment (dagar) [inline src/drivers/zenoh/CMakeLists.txt:47]: We cheat a bit with cmake initializing a submodule if the px4 module is included in a particular build.

``` 
px4_add_git_submodule(TARGET git_zenoh-pico PATH "zenoh-pico")
```

Then you can use `git_zenoh-pico` as a dependency below.

File: src/drivers/zenoh/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+unset(MESSAGE_QUIET)
+add_dependencies(zenohpico git_nuttx px4_platform)
```
Comment (dagar) [inline src/drivers/zenoh/CMakeLists.txt:49]: git_nuttx?

File: .gitmodules [hunk 3]
```diff
@@ -62,3 +62,13 @@
 	path = src/modules/uxrce_dds_client/Micro-XRCE-DDS-Client
 	url = https://github.com/PX4/Micro-XRCE-DDS-Client.git
 	branch = px4
+[submodule "src/modules/cdrstream/cyclonedds"]
+	path = src/lib/cdrstream/cyclonedds
+	url = https://github.com/eclipse-cyclonedds/cyclonedds
```
Comment (dagar) [inline .gitmodules:67]: Let's get these forked and carried under the PX4 org. In the past we've had submodules change out from under us in older releases and it was painful.

Comment (dagar) [reply .gitmodules:67]: @PetervdPerk-NXP here are the forks (you should have full access).

 - https://github.com/PX4/cyclonedds
 - https://github.com/PX4/rosidl
 - https://github.com/PX4/zenoh-pico

I usually keep the master/main branch in sync with the upstream repo, then create a `px4` branch so it's obvious.

**Comment by dagar** [comment]
> Hm to get Zenoh-pico to build, I think CMake 1.14 is mandatory need to check if CMake 1.13 might work as well.

We can re-evaluate the actual requirement if you're hitting it somewhere. Some of the older containers might just need a bump.

File: src/drivers/zenoh/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+target_compile_options(zenohpico PUBLIC -Wno-cast-align
+			-Wno-narrowing
+			-DUNIX_NO_MULTICAST_IF
+			-DZ_BATCH_SIZE_RX=512
+			-DZ_BATCH_SIZE_TX=512
+			-DZ_FRAG_MAX_SIZE=1024)
+
+px4_add_module(
+		MODULE drivers__zenoh
+		MAIN zenoh
```
[the remark this answers is not in the document]
Comment (PetervdPerk-NXP) [reply src/drivers/zenoh/CMakeLists.txt:59]: Based on uavcan & cyphal but I can move it modules that makes more sense.

[the remark this answers is not in the document]
Comment (dagar) [reply src/drivers/zenoh/CMakeLists.txt:59]: Although I'd like to change it, right now `drivers/uavcan` owns the actual CAN drivers for STM32, etc.

**Comment by Jaeyoung-Lim** [comment]
Yay!

**Comment by PetervdPerk-NXP** [comment]
These are the blocker for moving to CMake 3.14
| CI task               | Container                            | CMake version |
|-----------------------|--------------------------------------|---------------|
| mavros_mission_tests  | px4io/px4-dev-ros-melodic:2021-09-08 | 3.10.2        |
| mavros_offboard_tests | px4io/px4-dev-ros-melodic:2021-09-08 | 3.10.2        |
| compile_linux_arm64   | px4io/px4-dev-aarch64:2022-08-12     | 3.13.4        |
| clang-tidy            | px4io/px4-dev-clang:2021-09-08       | 3.10.2        |

**Comment by PetervdPerk-NXP** [comment]
Okay CI is successful now, (CMake 3.9 works and rosidl)

I've added 2 targets that can Zenoh and communicate with ROS2
```
nxp_mr-canhubk3_zenoh
px4_fmu-v6x_zenoh
```

Also made a `px4_sitl_zenoh`  target but it seems that sitl blocks file operation thus it fails when starting zenoh

**Comment by dirksavage88** [comment]
> Okay CI is successful now, (CMake 3.9 works and rosidl)
> 
> I've added 2 targets that can Zenoh and communicate with ROS2
> 
> ```
> nxp_mr-canhubk3_zenoh
> px4_fmu-v6x_zenoh
> px4_sitl_zenoh
> ```

I would be interested if this could work on the fmuk66 target as well since zenoh pico is supposed to be lightweight?

**Comment by PetervdPerk-NXP** [comment]
> I would be interested if this could work on the fmuk66 target as well since zenoh pico is supposed to be lightweight?

Albeit zenoh-pico is lightweight the current implementation might still be too big for the fmuk66.
Probably we could tune the profile by disabling multicast, tune the buffers and take a subset of uORB (currently we can convert all uORB messages to Zenoh and vice-versa).

Personally I would like to develop and improve zenoh-pico for mr-canhubk344 or other bigger targets before moving over to smaller targets.

File: Tools/msg/templates/zenoh/uorb_idl_header.h.em [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+from px_generate_uorb_topic_helper import * # this is in Tools/
+
+uorb_struct = '%s_s'%name_snake_case
+uorb_struct_upper = name_snake_case.upper()
+}@
+
+/****************************************************************
+
+  PX4 Cyclone DDS IDL to C Translator compatible idl struct
```
Comment (dagar) [inline Tools/msg/templates/zenoh/uorb_idl_header.h.em:13]: This template could be generic and default rather than zenoh specific.

Comment (PetervdPerk-NXP) [reply Tools/msg/templates/zenoh/uorb_idl_header.h.em:13]: I can restructure it, you okay with the following paths?
```
Tools/cdrstream/px_generate_cdrstream_files.py
Tools/cdrstream/templates
Tools/zenoh/px_generate_zenoh_files.py
Tools/zenoh/templates
```

**Review by dagar — APPROVED** [review]

**Comment by uupks** [comment]
Hi,  i'm facing the problem when i try to `make px4_sitl_zenoh`. 

```
[  7%] Generating px4/msg/ActionRequest.c, px4/msg/ActionRequest.h
[  8%] Built target cdr
idlc/bin/idlc: illegal option -- f cdrstream-desc
Usage: idlc [OPTIONS] FILE
Try 'idlc -h' for more information.
```

**Comment by PetervdPerk** [comment]
That errors seems to indicate there's a problem with a preinstalled version of `idlc` and then the cmake builds.

**Comment by uupks** [comment]
it seems like that  idlc doesn't support "-f cdrstream-desc" option

**Comment by PetervdPerk-NXP** [comment]
It does but you've got a "preinstalled" version. Are you by any chance running the px4 build after sourcing ros2?

**Comment by uupks** [comment]
i build px4_sitl_zenoh with a new shell environment , now it works..

Thanks !!

**Pull request merged**